### DISCUSSION TITLE
What are the best practices load large datasets of thousands of columns?

### DISCUSSION THREAD

**Discussion by fpixl** [duckdb/duckdb#8011] [Q&A]
Hi everyone,

I'm have a few large datasets I'm trying to load into DuckDB.

The datasets are in `.parquet` and follow a particular pattern:
- ~70% of the columns are string type
- ~25% are list of structs (between 2 or 3 keys)
- ~5% numeric columns

The dataset are partitioned and gets the number of rows for each fragment down to 2k rows each.
In general they follow this pattern:
- ~1k columns ~100k rows
- ~10k columns ~ 100k rows

To load to DuckDB I'm using pyarrow to read the dataset then loading each fragments into the table using `"INSERT INTO table SELECT * FROM 'file'"`.

For datasets with ~1k columns, each fragment takes around 50 seconds to load, and for ~10k columns is 7 minutes. 

I'd like to understand how to best deal with these kind of datasets, specially the ones with 10k columns which can take from multiple hours to days to load (depending on the number of rows).

How would you deal with such datasets?

**Comment by l1t1** [comment]
how about query on parquet files directly?

**Comment by fpixl** [comment]
I'm looking for query performance. Querying the parquet directly file is slower.

**Comment by Tishj** [comment]
> To load to DuckDB I'm using pyarrow to read the dataset then loading each fragments into the table using `"INSERT INTO table SELECT * FROM 'file'"`

I think l1t1 was talking about this, we have native support for reading parquet, you can skip this intermediate parquet -> arrow dataset step

Also, depending on your use case, you might want to disable insertion order preservation to improve performance
`SET preserve_insertion_order=false;`

**Reply by joaquimadraz** [reply]
My bad. I'm only loading a pyarrow dataset to get the fragments' path. I'm not loading a pyarrow table in memory.

The insert statements look something like `INSERT INTO table SELECT * FROM '/tmp/dataset/month=1/country=US/000-0.parquet'` for each fragment.

**Reply by Tishj** [reply]
In that case it looks like you want to use hive partitioning
These docs should help <https://duckdb.org/docs/data/partitioning/hive_partitioning> :)

**Reply by fpixl** [reply]
Hey @Tishj , thanks for this suggestion - really helped.

I'm now getting OOMs when loading the 10k column datasets. Maybe you can help.

> duckdb.OutOfMemoryException: Out of Memory Error: Failed to allocate block of 262144 bytes

Locally on my laptop (8 cores 32GB mem) I can load 10k columns with 1k rows and 10k columns with 10k rows. 10k columns with 100k rows OOMs
On a Fargate container on AWS, with the same specs (+200GB storage) it OOMs on 10k columns with 1k row.

It's odd that it's OOMing on AWS but not locally, is there any config I need to have in consideration when running this in a container?

How come datasets with fragments of 10MB and 50MB OOMs with 32GB of memory?

**Reply by fpixl** [reply]
Update: On AWS (16vCPU, 100GB of memory, 200GB storage) - 10k columns with 1k rows OOMs. These are fragments of 10M.

**Comment by joaquimadraz** [comment]
@Tishj Creating a table with hive partitioning is way faster! What's DuckDB doing that speeds up so much in comparison to loading each fragment?

**Reply by Tishj** [reply]
> @Tishj Creating a table with hive partitioning is way faster! What's DuckDB doing that speeds up so much in comparison to loading each fragment?

I'm not that schooled on the inner workings but due to the way the files and folders are organized and structured, that communicates a lot of information to us so we can parallelize and prtition the scans better
